### PR TITLE
feat: webhooks

### PR DESCRIPTION
**Pull request by Rajat-Ahuja1997**

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -100,20 +100,12 @@ repos:
       # Optional: Coverage check for main app only
       - id: coverage-main
         name: coverage check (main app)
-        entry: bash -c 'uv run coverage run -m pytest tests/ && uv run coverage report --fail-under=80'
+        entry: bash -c 'uv run coverage run -m pytest tests/ && uv run coverage report --fail-under=60'
```
Comment (Rajat-Ahuja1997) [inline .pre-commit-config.yaml]: we are at 61 on main

File: docs/v2/guides/mcp.mdx
Comment (VVoruganti) [inline docs/v2/guides/mcp.mdx:1]: Confused why there's a diff here. We already have these changes in main

File: src/config.py [hunk 2]
```diff
@@ -236,6 +236,14 @@ class SummarySettings(HonchoSettings):
     THINKING_BUDGET_TOKENS: Annotated[int, Field(default=512, gt=0, le=2000)] = 512
 
 
+class WebhookSettings(HonchoSettings):
+    model_config = SettingsConfigDict(env_prefix="WEBHOOKS_", extra="ignore")  # pyright: ignore
+
+    PROXY_URL: str | None = None
```
Comment (VVoruganti) [inline src/config.py]: Doesn't look like this used anywhere except a log statement

File: src/models.py [hunk 3]
```diff
@@ -28,6 +29,12 @@
 
 from .db import Base
 
+
+class WebhookStatus(str, Enum):
+    ENABLED = "enabled"
+    DISABLED = "disabled"
```
[the remark this answers is not in the document]
Comment (VVoruganti) [reply src/models.py]: +1 on this, looks unused

File: src/routers/webhooks.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+async def list_webhook_endpoints(
+    workspace_name: str | None = None,
+    jwt_params: JWTParams = Depends(require_auth()),
+    db: AsyncSession = db,
+):
+    """
+    List all webhook endpoints, optionally filtered by workspace.
+    """
+    # Handle workspace authorization if specified
+    if workspace_name and not jwt_params.ad and jwt_params.w != workspace_name:
```
Comment (VVoruganti) [inline src/routers/webhooks.py]: Double checking the logic here. 

If not an admin & also the workspace is not specified in the JWT. The JWT could just be an admin level without specifying the workspace so this would fail in that case? Shouldn't this be an `or` statement

File: src/routers/webhooks.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        raise AuthenticationException("Unauthorized access to resource")
+
+    return await crud.list_webhook_endpoints(db, workspace_name)
```
Comment (VVoruganti) [inline src/routers/webhooks.py]: might want to paginate this

File: src/routers/webhooks.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    Get or create a webhook endpoint URL.
+    """
+    if (
+        webhook.workspace_name
+        and not jwt_params.ad
+        and jwt_params.w != webhook.workspace_name
+    ):
+        raise AuthenticationException("Unauthorized access to resource")
+
+    return await crud.get_or_create_webhook_endpoint(db, webhook)
```
[the remark this answers is not in the document]
Comment (VVoruganti) [reply src/routers/webhooks.py]: For this it makes sense to have some common sense limit. I would say let's make a config value for WEBHOOK limit by workspace maybe start at 10.

File: src/utils/summarizer.py [hunk 7]
```diff
@@ -3,11 +3,11 @@
 import logging
 import time
 from enum import Enum
-from typing import TypedDict
 
 from mirascope import llm
 from sqlalchemy import update
 from sqlalchemy.ext.asyncio import AsyncSession
+from typing_extensions import TypedDict
```
[the remark this answers is not in the document]
Comment (VVoruganti) [reply src/utils/summarizer.py:10]: We should re-verify the minimum version of python that we're running on here. The `Dockerfile` is using 3.11 so should stick to that standard probably unless we have a strong reason to upgrade

File: src/webhooks/webhook_delivery.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+from src.config import settings
+from src.dependencies import tracked_db
+from src.webhooks.event_emitter import webhook_emitter
+from src.webhooks.events import WebhookEvent
+
+logger = logging.getLogger(__name__)
+
+
+class WebhookDeliveryService:
+    """Service for delivering webhook events to a proxy."""
```
Comment (VVoruganti) [inline src/webhooks/webhook_delivery.py]: Mentions a proxy again

File: src/webhooks/events.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    queue_type: str
+
+
+WebhookPayload = QueueEmptyPayload
+
+
+class WebhookEvent(BaseModel):
+    type: WebhookEventType
+    data: WebhookPayload
+    workspace_name: str
```
Comment (VVoruganti) [inline src/webhooks/events.py]: Create WebhookEvent Base `Pydantic` schema that other event types inherit from.

File: src/main.py [hunk 10]
```diff
@@ -101,8 +103,14 @@ def before_send(event: "Event", hint: "Hint") -> "Event | None":
 
 @asynccontextmanager
 async def lifespan(_: FastAPI):
-    yield
-    await engine.dispose()
+    # Start webhook delivery service
+    await webhook_delivery_service.start()
```
Comment (VVoruganti) [inline src/main.py]: Not sure about having the webhook service being a part of the API itself, how would events be communicated to it from the deriver or any other potential future services.

File: src/webhooks/event_emitter.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+            )
+            await self._event_queue.put(event)
+
+        except Exception as e:
+            logger.error(f"Failed to emit webhook event {event_type}: {e}")
+
+    async def emit_queue_empty(
+        self, workspace_name: str, payload: QueueEmptyPayload
+    ) -> None:
+        await self.emit_event(workspace_name, WebhookEventType.QUEUE_EMPTY, payload)
```
Comment (VVoruganti) [inline src/webhooks/event_emitter.py]: looking at the codebase don't actually see anywhere that events are emitted

File: src/webhooks/webhook_delivery.py [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+
+        Args:
+            payload: JSON string of the webhook payload
+
+        Returns:
+            HMAC-SHA256 signature as hex string
+        """
+        auth_secret = os.getenv("AUTH_JWT_SECRET")
+        if not auth_secret:
+            raise ValueError("AUTH_JWT_SECRET not found - cannot sign webhook")
```
Comment (VVoruganti) [inline src/webhooks/webhook_delivery.py]: Thinking about this, it would be valid to setup honcho without API auth, but with Webhooks in which we would need a new secret. Might make sense to split it up into a separate `WEBHOOK_SECRET` config value

File: src/config.py [hunk 13]
```diff
@@ -266,6 +274,7 @@ class AppSettings(HonchoSettings):
     DERIVER: DeriverSettings = Field(default_factory=DeriverSettings)
     DIALECTIC: DialecticSettings = Field(default_factory=DialecticSettings)
     SUMMARY: SummarySettings = Field(default_factory=SummarySettings)
+    WEBHOOKS: WebhookSettings = Field(default_factory=WebhookSettings)
```
Comment (VVoruganti) [inline src/config.py]: need to add this to the section map at the top so it works properly with toml files as well

File: src/crud/webhook.py [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+        await get_workspace(db, webhook.workspace_name)
+
+    # Check if endpoint already exists
+    stmt = select(models.WebhookEndpoint).where(
+        models.WebhookEndpoint.url == webhook.url,
+    )
+    if webhook.workspace_name:
+        stmt = stmt.where(
+            models.WebhookEndpoint.workspace_name == webhook.workspace_name
+        )
```
Comment (VVoruganti) [inline src/crud/webhook.py]: I think this optional workspace scoping is just creating weird behavior. You could have duplicate urls with one being workspace scoped and another not? And then it would send duplicates everywhere. 

The more I'm looking at this the more it makes sense to just enforce workspace level webhooks

File: src/crud/webhook.py [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+    await db.commit()
+    await db.refresh(webhook_endpoint)
+
+    logger.info(f"Webhook endpoint created: {webhook.url}")
+    return schemas.WebhookEndpoint.model_validate(webhook_endpoint)
+
+
+async def list_webhook_endpoints(
+    db: AsyncSession, workspace_name: str | None = None
+) -> list[schemas.WebhookEndpoint]:
```
Comment (VVoruganti) [inline src/crud/webhook.py]: Should paginate for lists

File: src/webhooks/webhook_delivery.py (hunk 12 above)
Comment (Rajat-Ahuja1997) [reply src/webhooks/webhook_delivery.py]: added -> going to add a commit in the groudon PR to set the `WEBHOOK_SECRET` when creating the machine

File: src/config.py [hunk 16]
```diff
@@ -236,6 +237,13 @@ class SummarySettings(HonchoSettings):
     THINKING_BUDGET_TOKENS: Annotated[int, Field(default=512, gt=0, le=2000)] = 512
 
 
+class WebhookSettings(HonchoSettings):
+    model_config = SettingsConfigDict(env_prefix="WEBHOOKS_", extra="ignore")  # pyright: ignore
+
+    WEBHOOK_SECRET: str | None = None  # Must be set if configuring webhooks
+    WORKSPACE_LIMIT: int = 10
+
```
[the remark this answers is not in the document]
Comment (Rajat-Ahuja1997) [reply src/config.py]: making it WEBHOOKS.MAX_WORKSPACE_LIMIT

File: src/deriver/consumer.py [hunk 17]
```diff
@@ -1,11 +1,12 @@
 import logging
 from typing import Any
 
+import httpx
 from pydantic import ValidationError
 from rich.console import Console
 
 from .deriver import Deriver
-from .queue_payload import RepresentationPayload, SummaryPayload
+from .queue_payload import RepresentationPayload, SummaryPayload, WebhookQueuePayload
```
Comment (VVoruganti) [inline src/deriver/consumer.py]: Should be consistent in naming either all be just `XYZPayload` or `XYZQueuePayload`

File: src/deriver/consumer.py [hunk 18]
```diff
@@ -15,25 +16,40 @@
 deriver = Deriver()
 
 
-async def process_item(payload: dict[str, Any]) -> None:
+async def process_item(
+    client: httpx.AsyncClient, task_type: str, payload: dict[str, Any]
```
Comment (VVoruganti) [inline src/deriver/consumer.py]: a bit skeptical that we need to be passing a `httpx` client around, can't we just make it on the stop or use the `httpx` library directly. I was under the impression that you only need to do this when you want to create a custom client with pre-set flags.

File: src/deriver/queue_manager.py [hunk 19]
```diff
@@ -2,15 +2,16 @@
 import signal
 from asyncio import Task
 from collections.abc import Sequence
-from dataclasses import dataclass
-from datetime import UTC, datetime, timedelta
+from datetime import datetime, timedelta, timezone
 from logging import getLogger
+from sqlite3 import IntegrityError
```
Comment (VVoruganti) [inline src/deriver/queue_manager.py]: Shouldn't this be coming from sqlalchemy not sqlite3? We aren't using sqlite anywhere

File: src/deriver/utils.py [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+        session_id = payload["session_name"] or "None"
+        return f"{task_type}:{session_id}:{sender_name}:{target_name}"
+
+    if task_type == "webhook":
+        workspace_name = payload["workspace_name"]
+        if not workspace_name:
+            raise ValueError("Workspace name is required for webhook tasks")
+        return f"webhook:{workspace_name}"
+
+    raise ValueError(f"Invalid task type: {task_type}")
```
Comment (VVoruganti) [inline src/deriver/utils.py]: Have we done any testing to see how this behavior looks in a normal app. Mainly is there fear of the workers being held up with webhooks, or since there's only ever at most one worker handling webhooks for a workspace it should be ok?

File: src/models.py [hunk 21]
```diff
@@ -83,6 +83,7 @@ class Workspace(Base):
     id: Mapped[str] = mapped_column(TEXT, default=generate_nanoid, primary_key=True)
     name: Mapped[str] = mapped_column(TEXT, index=True, unique=True)
     peers = relationship("Peer", back_populates="workspace")
+    webhook_endpoints = relationship("WebhookEndpoint", back_populates="workspace")
```
Comment (VVoruganti) [inline src/models.py:86]: This is the one table that is capitalized, should be consistent

Comment (VVoruganti) [reply src/models.py:86]: should that value be the class name or the table name?

File: src/models.py [hunk 22]
```diff
@@ -376,26 +383,36 @@ class ActiveQueueSession(Base):
...
+    id: Mapped[str] = mapped_column(TEXT, default=generate_nanoid, primary_key=True)
+    workspace_name: Mapped[str] = mapped_column(
+        ForeignKey("workspaces.name"), index=True, nullable=False
+    )
+    url: Mapped[str] = mapped_column(TEXT, nullable=False)
+    created_at: Mapped[datetime.datetime] = mapped_column(
+        DateTime(timezone=True), default=func.now()
+    )
+
+    workspace = relationship("Workspace", back_populates="webhook_endpoints")
```
Comment (VVoruganti) [inline src/models.py:406]: Same concern about capitalization

File: src/routers/webhooks.py [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+async def delete_webhook_endpoint(
+    endpoint_id: Annotated[str, Path(description="Webhook endpoint ID")],
+    jwt_params: JWTParams = Depends(require_auth()),
+    db: AsyncSession = db,
+) -> None:
+    """
+    Delete a specific webhook endpoint.
+    """
+
+    if not jwt_params.ad:
```
Comment (VVoruganti) [inline src/routers/webhooks.py]: I think a workspace level key should be able to make these requests too

File: src/webhooks/event_publisher.py [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+        except Exception as e:
+            logger.exception(
+                "Failed to publish webhook event %s: %s",
+                event.type,
+                e,
+            )
+
+
+# Global event publisher instance
+webhook_publisher = WebhookQueuePublisher()
```
Comment (VVoruganti) [inline src/webhooks/event_publisher.py]: Is there a reason to make this a class, could just be a method we call directly so we can avoid having a singleton that needs to be managed in memory.

File: src/webhooks/events.py [hunk 25]
```diff
@@ -0,0 +1,24 @@
...
+    """Base class for all webhook events."""
+
+    workspace_name: str
+
+
+class QueueEmptyEvent(BaseWebhookEvent):
+    """Webhook event for when a queue becomes empty."""
+
+    type: WebhookEventType = WebhookEventType.QUEUE_EMPTY
+    queue_type: str
```
Comment (VVoruganti) [inline src/webhooks/events.py:30]: Should make this Literals as well just for safety.

File: src/webhooks/webhook_delivery.py [hunk 26]
```diff
@@ -0,0 +1,105 @@
...
+                    },
+                )
+                if 200 <= response.status_code < 300:
+                    logger.info(
+                        f"Successfully delivered webhook {payload.event_type} to {url}"
+                    )
+                else:
+                    logger.error(
+                        f"Failed delivery for {payload.event_type} to {url}. Status: {response.status_code}"
+                    )
```
Comment (VVoruganti) [inline src/webhooks/webhook_delivery.py]: Should be able to make these in parallel or in batches

File: src/webhooks/webhook_delivery.py [hunk 27]
```diff
@@ -0,0 +1,105 @@
...
+from sqlalchemy.ext.asyncio import AsyncSession
+
+from src.config import settings
+from src.deriver.queue_payload import WebhookQueuePayload
+
+logger = logging.getLogger(__name__)
+
+
+class WebhookDeliveryService:
+    """Service for delivering webhook events from a queue payload."""
```
Comment (VVoruganti) [inline src/webhooks/webhook_delivery.py]: Could move the custom `httpx` client into this class if necessary to keep it

File: tests/routes/test_webhooks.py
Comment (VVoruganti) [inline tests/routes/test_webhooks.py:1]: Add tests for creating more endpoints than the max and for creating duplicate endpoints. Also check that you can add the same endpoint on two different workspaces without error

File: src/webhooks/event_publisher.py [hunk 28]
```diff
@@ -0,0 +1,56 @@
...
+from src.models import QueueItem
+from src.webhooks.events import WebhookEvent
+
+logger = logging.getLogger(__name__)
+
+
+class WebhookQueuePublisher:
+    """Places webhook events onto the database queue."""
+
+    async def publish_event(self, event: WebhookEvent) -> None:
```
Comment (VVoruganti) [inline src/webhooks/event_publisher.py]: Don't see anything actually using publish_event, can we also implement the emitting the QueueEmpty event

File: migrations/versions/88b0fb10906f_add_webhooks_table.py [hunk 29]
```diff
@@ -0,0 +1,133 @@
...
+        "queue",
+        sa.Column("task_type", sa.TEXT(), nullable=True),
+        schema=schema,
+    )
+
+    op.add_column(
+        "queue",
+        sa.Column("work_unit_key", sa.Text(), nullable=True),
+        schema=schema,
+    )
```
Comment (VVoruganti) [inline migrations/versions/88b0fb10906f_add_webhooks_table.py:77]: Should add a data transformation step that'll fill up the `work_unit_key` column for old columns, otherwise this upgrade would break any existing unprocessed queue items.

**Review by VVoruganti — CHANGES_REQUESTED** [review]
Overall structure looks good a couple of nitpicks on code style and testing. Also implement using the webhooks. 

Most important stuff
- Add more tests
- Implement Queue Empty
- Make the alembic migration more safe

File: src/webhooks/events.py
Comment (VVoruganti) [inline src/webhooks/events.py:1]: Createa a `TESTEventType` and an endpoint that let's devs emit a test webhook.

File: src/deriver/queue_manager.py (hunk 19 above)
Comment (Rajat-Ahuja1997) [reply src/deriver/queue_manager.py]: good catch

File: src/webhooks/event_publisher.py (hunk 24 above)
Comment (Rajat-Ahuja1997) [reply src/webhooks/event_publisher.py]: made this and the delivery service use a function

File: src/webhooks/webhook_delivery.py (hunk 26 above)
Comment (Rajat-Ahuja1997) [reply src/webhooks/webhook_delivery.py]: added basic parallelization for now

File: src/webhooks/webhook_delivery.py (hunk 27 above)
Comment (Rajat-Ahuja1997) [reply src/webhooks/webhook_delivery.py]: removed

File: src/deriver/consumer.py (hunk 18 above)
Comment (Rajat-Ahuja1997) [reply src/deriver/consumer.py]: removed

File: src/deriver/utils.py (hunk 20 above)
Comment (Rajat-Ahuja1997) [reply src/deriver/utils.py]: i think it should be fine considering we only have at most one worker handling webhooks for a workspace. i'm more concerned with a single worker processing webhooks slowly if a workspace has a lot of webhooks, but we can configure a webhook+event_type worker work unit if needed in the future

Comment (Rajat-Ahuja1997) [reply src/deriver/utils.py]: i tested this with some webhook calls (test endpoint) along with basic representation deriving and it seemed fine but didn't extensively stress test. but with our current pattern i don't think workers will be held up with webhooks unless a tenant has several workspaces that are all firing tons of webhooks. even then, webhooks should process very quickly

File: src/routers/webhooks.py (hunk 23 above)
Comment (Rajat-Ahuja1997) [reply src/routers/webhooks.py]: fixed

File: src/models.py (hunk 21 above)
Comment (Rajat-Ahuja1997) [reply src/models.py:86]: discussed offline, this is good

File: src/deriver/queue_manager.py [hunk 30]
```diff
@@ -145,13 +116,13 @@ async def cleanup(self):
...
+    async def get_available_work_units(self, db: AsyncSession) -> Sequence[str]:
         """
         Get available work units that aren't being processed.
-        Returns a list of WorkUnit objects.
+        Returns a list of work unit keys.
         """
         # Clean up stale work units
-        five_minutes_ago = datetime.now(UTC) - timedelta(
+        five_minutes_ago = datetime.now(timezone.utc) - timedelta(
             minutes=settings.DERIVER.STALE_SESSION_TIMEOUT_MINUTES
```
Comment (Rajat-Ahuja1997) [inline src/deriver/queue_manager.py:126]: this timeout is currently 5min in testing/prod. we should conside adding an additional check to the stale cleanup query e.g. 
```
.where(
    ~exists(
        select(1)
        .select_from(models.QueueItem)
        .where(
            models.QueueItem.work_unit_key
            == models.ActiveQueueSession.work_unit_key,
            ~models.QueueItem.processed,
        )
    )
)
```
and then we can have an additional check where we delete any active queue sessions where last_update >20min (likely a crashed worker)

File: migrations/versions/88b0fb10906f_add_webhooks_table.py [hunk 31]
```diff
@@ -0,0 +1,175 @@
...
+                        COALESCE(s.workspace_name, q.payload->>'workspace_name', 'None') || ':' ||
+                        COALESCE(q.payload->>'session_name', 'None') || ':' ||
+                        COALESCE(q.payload->>'sender_name', 'None') || ':' ||
+                        COALESCE(q.payload->>'target_name', 'None')
+                END
+            FROM {schema}.sessions s
+            WHERE s.id = q.session_id AND q.processed = false AND q.work_unit_key IS NULL
+            """
+        )
+    )
```
Comment (VVoruganti) [inline migrations/versions/88b0fb10906f_add_webhooks_table.py:109]: Make the queue work_unit column nullable after the transformation

Comment (VVoruganti) [reply migrations/versions/88b0fb10906f_add_webhooks_table.py:109]: same thing with the task_type column

File: src/deriver/utils.py [hunk 32]
```diff
@@ -0,0 +1,67 @@
...
+
+
+class ParsedWorkUnit(TypedDict):
+    task_type: str
+    workspace_name: str
+    session_name: str | None
+    sender_name: str | None
+    target_name: str | None
+
+
```
Comment (VVoruganti) [inline src/deriver/utils.py:11]: Potential optimization for the future is to put this all into a pydantic object that can take in the work_unit_key and parse it automatically as well as handle the transform back and forth

Comment (Rajat-Ahuja1997) [reply src/deriver/utils.py:11]: agree

**Review by VVoruganti — APPROVED** [review]
